docs/cookbook/prompt_engineering/chaining_based/self_refine.md

```diff
@@ -0,0 +1,173 @@
+# Self-Refine: Enhancing LLM Outputs Through Iterative Self-Improvement
+
+This recipe demonstrates how to implement the Self-Refine technique using Large Language Models (LLMs) with Mirascope. Self-Refine is a prompt engineering method that enhances an LLM's output by iteratively generating feedback and improving its responses.
+
+??? tip "Mirascope Concepts Used"
+
+    - [Prompts](../../../learn/prompts.md)
+    - [Calls](../../../learn/calls.md)
+    - [Dynamic Configuration](../../../learn/dynamic_configuration.md)
+
+!!! note "Background"
+
+    [Self-refine](https://arxiv.org/pdf/2303.17651) is a prompt engineering technique where a model gives feedback about its answer and uses the feedback to generate a new answer. This self refinement can take place multiple times to generate the final answer. Self-refine is helpful for reasoning, coding, and generation tasks.
+
+## Basic Self-Refine Implementation
+
+Let's start with a basic implementation of Self-Refine using Mirascope:
+
+```python
+from mirascope.core import openai, prompt_template
+
+@openai.call(model="gpt-4o-mini")
+@prompt_template("""{query}""")
+def call(query: str): ...
+
+@openai.call(model="gpt-4o-mini")
+@prompt_template(
+    """
+    Here is a query and a response to the query. Give feedback about the answer,
+    noting what was correct and incorrect.
+    Query:
+    {query}
+    Response:
+    {response}
+    """
+)
+def evaluate_response(query: str, response: str): ...
+
+@openai.call(model="gpt-4o-mini")
+@prompt_template(
+    """
+    For this query:
+    {query}
+    The following response was given:
+    {response}
+    Here is some feedback about the response:
+    {feedback}
+
+    Consider the feedback to generate a new response to the query.
+    """
+)
+def generate_new_response(query: str, response: str) -> openai.OpenAIDynamicConfig:
+    feedback = evaluate_response(query, response)
+    print(feedback)
+    return {"computed_fields": {"feedback": feedback}}
+
+def self_refine(query: str, depth: int) -> str:
+    response = call(query).content
```

Comment:
you should pass the original response down the chain so computed fields retain the original response